compair/static/modules/image-viewer/image-viewer-directive.js

```diff
@@ -0,0 +1,273 @@
+                imgFile: '<?',      // ... or File / Blob object...
+                orientation: '=?',  // Orientation of current view
+                uploadFile: '=?',   // File object of original image oriented as current view
+                options: '<?',
+                onLoadCallback: '&?'   // callback when image loaded
+            },
+            template: template,
+            link: function(scope, element, attributes) {
+                scope.canvasLoaded = false;
+
+                // default values
+                _.defaults(scope.orientation, { rotate: 0, zoom: 1 });
+                scope.options = $.extend({
+                    backgroundColor: 'transparent',
+                }, scope.options);
+
+                // round to nearest 90
+                scope.orientation.rotate = Math.round(scope.orientation.rotate / 90.0) * 90;
+                if ((scope.orientation.rotate <= -360) || (scope.orientation.rotate >= 360)) {
+                    scope.orientation.rotate = 0;
+                }
+
+                var canvas = element.find('canvas')[0];
+                var ctx = canvas.getContext("2d");
+                var picPos = { x : 0, y : 0 };  // origin of the image
+
+                var uploadFileMimeType = scope.imgFile? scope.imgFile.type : 'image/png';
+                var img = new Image();
+                img.onload = onLoadImage;
+
+                // load image either by src URL or by selected file
+                if (scope.src && typeof(scope.src) === 'string') {
+                    img.src = scope.src;
+                } else if (scope.imgFile) {
+                    var reader = new FileReader();
+                    reader.onload = onLoadFile;
+                    reader.readAsDataURL(scope.imgFile);
+                }
+
+                // for monitoring canvas resize
+                function parentChange() {
+                    return {
+                        height: canvas.parentNode.clientHeight,
+                        width: canvas.parentNode.clientWidth };
+                }
+
+                function onLoadFile(event) {
+                    img.src = event.target.result;
+                }
+
+                function onLoadImage() {
+                    scope.canvasLoaded = true;
+
+                    // TODO: parse EXIF data to auto rotate the image
+                    resizeCanvas();
+
+                    // start monitoring resize
+                    scope.$watch(parentChange, function() {
+                        resizeCanvas();
+                    }, true);
+
+                    refreshUploadFile().then(function() {
+                        if (scope.onLoadCallback) {
+                            scope.onLoadCallback();
+                        }
+                    });
+                }
+
+                // https://stackoverflow.com/questions/4998908/convert-data-uri-to-file-then-append-to-formdata
+                function dataUrlToBlob(dataUrl, mimeType) {
```

Comment:
why is `mimeType` passed around throughout the functions instead of using `uploadFileMimeType` directly (mimetype shouldn't change unless I'm missing something)?